### PR TITLE
feat(cli): allow users to download models from Kaggle

### PR DESCRIPTION
**Pull request by KeijiBranshi**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

https://github.com/pytorch/torchtune/issues/1852

#### Changelog
What are the changes made in this PR?
* 5527fd88a193811fd2c471222be582a0b508e203
  * Added 3 new command line arguments: `--source`, `--kaggle-username`, and `--kaggle-api-key`
  * Splits download logic based on `--source` value (only valid options are `huggingface` and `kaggle`). When `--source` is _not_ provided, defaults to `huggingface`
  * Kaggle download logic placed in helper function `_download_from_kaggle()`
  * The new logic _mostly_ mirrors `_download_from_huggingface()` with a few notable exceptions:
    * Authentication with `kagglehub` requires the caller's username and api token as separate arguments to `kagglehub.set_kaggle_credentials()` (these can also be set as environment variables). Since both are required, appropriate error messages are raised when one is omitted. However, since Kaggle supports anonymous downloads on public non-gated models, the option to omit both is also supported.
    * The exceptions thrown by `kagglehub.model_download` are different than those raised by `huggingface_hub.snapshot_download`. 
    * `kagglehub.model_download` doesn't support direct download to a caller-specified directory via `--output-dir`. This feature is being worked on in `kagglehub` (see https://github.com/Kaggle/kagglehub/issues/179), but in the meantime a warning was just placed here instead.
    * some extra validation is done on Kaggle model handles (see `_validate_kaggle_model_handle`) as per the requests made [here](https://github.com/pytorch/torchtune/issues/1852#issuecomment-2443966531). This validation is primarily used to raise warnings when downloads may not be guaranteed to work with torchtune.
* a3776d6a7c0d46b1efe5f723e832feb8681a2f57
    * tests were added to validate the CLI behavior when `--source kaggle` is provided
* b30f741f79c8901ffffd6a02d53e81051a4ae32f
    * updated documentation

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [x] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [ ] ~~run recipe tests via `pytest tests -m integration_test`~~
  - [x] screenshot as per the instructions in [this comment](https://github.com/pytorch/torchtune/issues/1852#issuecomment-2454478446): 8zC8bxfkWnas3ex
- [ ] ~~manually run any new or modified recipes with sufficient proof of correctness~~
- [ ] ~~include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)~~

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [x] I have added an example to docs or docstrings

### CODE REVIEW

**Comment by joecummings** [comment]
Awesome stuff!! Will be reviewing this today or tomorrow :)

File: docs/source/tune_cli.rst
Comment (KeijiBranshi) [inline docs/source/tune_cli.rst:53]: just noticed that `--ignore-patterns` should probably emit a warning too for Kaggle models. It has a default value though (`"*.safetensors"`), which should probably be kept for backwards compatibility. 

https://github.com/pytorch/torchtune/blob/1eb7785321dcc651e5bf5b44842f790b7b873aa5/torchtune/_cli/download.py#L90-L97

So we'd need a way to distinguish when a user explicitly sets it 🤔

**Review by joecummings — COMMENTED** [review]
A few comments/questions mostly about authentication, but overall this looks great!

File: docs/source/tune_cli.rst
Comment (KeijiBranshi) [reply docs/source/tune_cli.rst:57]: I wasn't sure if this would add more confusion for the developer experience, considering `kagglehub.login()` is only invoked from Python but `tune download` is a CLI. Granted, notebooks incorporate both. I'm curious for your thoughts on that though.

File: torchtune/_cli/download.py
Comment (KeijiBranshi) [reply torchtune/_cli/download.py:130]: Can do!

side note: `kagglehub.model_download` will [automatically grab those credentials](https://github.com/Kaggle/kagglehub/blob/8b1fae8632f9d381cebb14ec50c56d7ff5fbeb1b/src/kagglehub/clients.py#L94) from the environment, so there's technically no need to set it ourselves here. But I'll still add it to help be more explicit in the scope of `download.py`.

File: docs/source/tune_cli.rst
Comment (KeijiBranshi) [reply docs/source/tune_cli.rst:53]: Yep. For example, our Transformers variation of Gemma mirrors the files on the Hugging Face page:

https://www.kaggle.com/models/google/gemma/transformers

File: docs/source/tune_cli.rst
Comment (KeijiBranshi) [reply docs/source/tune_cli.rst:57]: After more thought, I'm not sure `kagglehub.login()` is viable here. It will only set the credentials within the running Python context, and will not modify the environment variables to authenticate across contexts. So even if a user calls in somewhere in their jupyter/kaggle notebook, those credentials will not carry over to the shell command `tune download`.

cc @rosbo for your take here

Comment (rosbo) [reply docs/source/tune_cli.rst:57]: That's correct, `kagglehub.login()` doesn't persist the credentials on "disk".

File: docs/source/tune_cli.rst
Comment (joecummings) [reply docs/source/tune_cli.rst:53]: Ah I see. Yeah we should log a warning about this too for Kaggle and honestly, we should just get rid of this default altogether IMO. When we were bootstrapping up, we supported only Meta model formats to make things easier and so didn't want to download double weights, but now the majority of our models are safetensor format.

File: docs/source/tune_cli.rst
Comment (joecummings) [reply docs/source/tune_cli.rst:57]: Got it - so the user would have to explicitly set their credentials as environment variables with e.g. `export KAGGLE_TOKEN=asdkfasdfasdfasdf`, is that correct?

Comment (joecummings) [reply docs/source/tune_cli.rst:57]: The reason I'm asking is that the username is easy to remember and pass in again and again, but the token is only visible once and can be sensitive to just store in a note. I might prefer that we add a note recommending they save it to their `.bash_profile` or something.

File: torchtune/_cli/download.py
Comment (joecummings) [inline torchtune/_cli/download.py:224]: Could we check to see if the API key is defined in an environment variable here and allow it to pass through if so?

File: torchtune/_cli/download.py
Comment (joecummings) [inline torchtune/_cli/download.py:228]: Same comment here about kaggle username

File: docs/source/tune_cli.rst
Comment (KeijiBranshi) [reply docs/source/tune_cli.rst:57]: > so the user would have to explicitly set their credentials as environment variables with e.g. `export KAGGLE_TOKEN=asdkfasdfasdfasdf`

yes, that's correct (with a small note - the necessary env vars are `KAGGLE_KEY` **and** `KAGGLE_USERNAME`).

> but the token is only visible once and can be sensitive to just store in a note

agreed. I think it's safer to push users towards using environment variables. In this case, I was mostly going for consistency with the `--hf-token` argument.

Comment (KeijiBranshi) [reply docs/source/tune_cli.rst:57]: just to make sure I understand the immediate suggestion. should we:

1. completely _remove_ the `--kaggle-username` and `--kaggle-api-key` arguments, and/or 2. opt for an error message to set the env vars in the users shell profile

Comment (KeijiBranshi) [reply docs/source/tune_cli.rst:57]: also, another alternative to consider: we could add a `--kaggle-login` argument instead. when present, `torchtune` would call `kagglehub.login()` before running `kagglehub.model_download()`.

File: torchtune/_cli/download.py
Comment (KeijiBranshi) [reply torchtune/_cli/download.py:224]: Can do! 

passing, non-blocking comment: I initially opted not to do something like this since the correct env var to use is a bit of a hidden global implementation detail in `kagglehub`. Wasn't sure if it would feel a little mysterious just reading it from `torchtune`'s context.

File: docs/source/tune_cli.rst
Comment (joecummings) [reply docs/source/tune_cli.rst:57]: > just to make sure I understand the immediate suggestion. should we:
> 
> 1. completely _remove_ the `--kaggle-username` and `--kaggle-api-key` arguments, and/or
> 2. opt for an error message to set the env vars in the users shell profile

Keep the username and api key arguments. And if they are NOT empty, log a "tip" that the user can store these in their environment instead of passing these arguments in every time.

File: docs/source/tune_cli.rst
Comment (KeijiBranshi) [reply docs/source/tune_cli.rst:53]: Addressed here: f91607bf813153d16c0d4f308c70b32fbb4e2bd9

File: torchtune/_cli/download.py
Comment (KeijiBranshi) [reply torchtune/_cli/download.py]: resolved here: 1cd0de57cd405c40d8b72284cbc8f472123e07f2

File: torchtune/_cli/download.py
Comment (KeijiBranshi) [reply torchtune/_cli/download.py:130]: Resolved here: c2dee3a6911a48fa718fb40face90aa190cabb9b

File: docs/source/tune_cli.rst
Comment (KeijiBranshi) [reply docs/source/tune_cli.rst:57]: Resolved here: c2dee3a6911a48fa718fb40face90aa190cabb9b

File: torchtune/_cli/download.py
Comment (KeijiBranshi) [reply torchtune/_cli/download.py]: Resolved here: c2dee3a6911a48fa718fb40face90aa190cabb9b

File: torchtune/_cli/download.py
Comment (KeijiBranshi) [reply torchtune/_cli/download.py]: Resolved here: c2dee3a6911a48fa718fb40face90aa190cabb9b

File: torchtune/_cli/download.py
Comment (KeijiBranshi) [reply torchtune/_cli/download.py:130]: Closing this thread out to consolidate it into https://github.com/pytorch/torchtune/pull/2002#discussion_r1846913699

**Review by joecummings — APPROVED** [review]
Will wait for CI, but LGTM!

Thanks @KeijiBranshi for your hard work on this and making it easier for torchtune users to download models from the Kaggle Hub!

**Comment by KeijiBranshi** [comment]
> Will wait for CI, but LGTM!
> 
> Thanks @KeijiBranshi for your hard work on this and making it easier for torchtune users to download models from the Kaggle Hub!

Thanks so much for your help and review! Really appreciated your guidance on everything.

Looks like some of my test are failing. Could've sworn I ran before pushing, but i guess it slipped through. Resolving those now 🙂

**Comment by KeijiBranshi** [comment]
@joecummings success!

Thanks again for all your help. Let me know if anything else needs to be done for the release process 🚀

**Comment by KeijiBranshi** [comment]
@joecummings quick fyi, I just rebased to the most recent `main` branch commit. 

Let us know if there's anything else we should do to get this merged 🙂

**Comment by joecummings** [comment]
> @joecummings quick fyi, I just rebased to the most recent `main` branch commit.
> 
> Let us know if there's anything else we should do to get this merged 🙂

Thanks for the rebase. Will merge this today 💯

**Comment by KeijiBranshi** [comment]
> > @joecummings quick fyi, I just rebased to the most recent `main` branch commit.
> > Let us know if there's anything else we should do to get this merged 🙂
> 
> Thanks for the rebase. Will merge this today!! Afterwards, I'll kick off a "nightly" build so users can have immediate access to this feature even before tomorrow's build of torchtune.

Thank you so much!! Looking forward to it. 

Please ping me if things break or users see issues 🙂. If I'm unresponsive for any reason, you can also email kaggle-models@google.com.

**Pull request merged**
